File: apps/api/app/schemas/ai_accountant.py
from datetime import date
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.accounting import LedgerCategory, VoucherType
from app.schemas.accounting import VoucherResponse


class NaturalLanguageEntryRequest(BaseModel):
    text: str = Field(min_length=3, max_length=800)
    transaction_date: date | None = None
    language: str = "auto"


class SuggestedLedger(BaseModel):
    ledger_id: UUID | None
    ledger_name: str
    category: LedgerCategory
    reason: str
    should_create: bool = False


class SuggestedPostingLine(BaseModel):
    ledger_id: UUID | None
    ledger_name: str
    debit: Decimal
    credit: Decimal
    reason: str


class AiVoucherSuggestionResponse(BaseModel):
    suggestion_id: UUID
    input_text: str
    voucher_type: VoucherType
    voucher_date: date
    amount: Decimal
    confidence: Decimal
    gst_applicable: bool
    suggested_gst_rate: Decimal | None
    suggested_ledgers: list[SuggestedLedger]
    lines: list[SuggestedPostingLine]
    explanation: str
    validation_errors: list[str]
    can_post: bool
    model_name: str


class ConfirmAiPostingRequest(BaseModel):
    suggestion_id: UUID


class RejectAiSuggestionRequest(BaseModel):
    reason: str = Field(min_length=2, max_length=300)


class ConfirmAiPostingResponse(BaseModel):
    suggestion_id: UUID
    voucher: VoucherResponse

File: apps/web/lib/api/ai-accountant.ts
import { publicEnv } from "@/lib/config";
import { safeApiErrorMessage } from "@/lib/api/safe-error";
import type { LedgerCategory, VoucherType, Voucher } from "@/lib/api/accounting";

export type AiSuggestion = {
  suggestion_id: string;
  input_text: string;
  voucher_type: VoucherType;
  voucher_date: string;
  amount: string;
  confidence: string;
  gst_applicable: boolean;
  suggested_gst_rate: string | null;
  suggested_ledgers: Array<{
    ledger_id: string | null;
    ledger_name: string;
    category: LedgerCategory;
    reason: string;
    should_create: boolean;
  }>;
  lines: Array<{
    ledger_id: string | null;
    ledger_name: string;
    debit: string;
    credit: string;
    reason: string;
  }>;
  explanation: string;
  validation_errors: string[];
  can_post: boolean;
  model_name: string;
};

export type ConfirmAiPostingResponse = {
  suggestion_id: string;
  voucher: Voucher;
};

async function api<T>(path: string, token: string, body: unknown): Promise<T> {
  const response = await fetch(`${publicEnv.NEXT_PUBLIC_API_URL}${path}`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json"
    },
    body: JSON.stringify(body)
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: "Request failed" }));
    throw new Error(safeApiErrorMessage(error.detail, "AI accountant request could not be completed. Please try again."));
  }

  return response.json() as Promise<T>;
}

export const aiAccountantApi = {
  parse: (companyId: string, token: string, text: string) =>
    api<AiSuggestion>(`/companies/${companyId}/ai-accountant/parse`, token, { text }),
  suggestVoucher: (companyId: string, token: string, text: string) =>
    api<AiSuggestion>(`/companies/${companyId}/ai-accountant/suggest-voucher`, token, { text }),
  confirm: (companyId: string, token: string, suggestionId: string) =>
    api<ConfirmAiPostingResponse>(`/companies/${companyId}/ai-accountant/confirm`, token, {
      suggestion_id: suggestionId
    }),
  reject: (companyId: string, token: string, suggestionId: string, reason: string) =>
    api<AiSuggestion>(`/companies/${companyId}/ai-accountant/suggestions/${suggestionId}/reject`, token, {
      reason
    })
};
